File: Hangman.java
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class Hangman {

    public static String getRandomWord() {
        String[] words = {"example", "hangman", "computer", "programming", "cplusplus", "random", "word"};
        Random random = new Random();
        int randomIndex = random.nextInt(words.length);
        return words[randomIndex];
    }

    public static boolean isAlreadyGuessed(ArrayList<Character> guessedLetters, char letter) {
        for (char guessedLetter : guessedLetters) {
            if (guessedLetter == letter) {
                return true;
            }
        }
        return false;
    }

    public static class Word {
        private String hiddenWord;
        private StringBuilder guessedWord;
        private ArrayList<Character> guessedLetters;

        public Word(String hiddenWord) {
            this.hiddenWord = hiddenWord;
            this.guessedWord = new StringBuilder("_".repeat(hiddenWord.length()));
            this.guessedLetters = new ArrayList<>();
        }

        public boolean guessLetter(char letter) {
            if (isAlreadyGuessed(guessedLetters, letter)) {
                return false;
            }
            guessedLetters.add(letter);

            boolean correctGuess = false;
            for (int i = 0; i < hiddenWord.length(); i++) {
                if (hiddenWord.charAt(i) == letter) {
                    guessedWord.setCharAt(i, letter);
                    correctGuess = true;
                }
            }
            return correctGuess;
        }

        public String getGuessedWord() {
            return guessedWord.toString();
        }

        public String getHiddenWord() {
            return hiddenWord;
        }

        public boolean isFullyGuessed() {
            return guessedWord.toString().equals(hiddenWord);
        }
    }

    public static class HangmanGame {
        private Word word;
        private int maxGuesses;
        private int incorrectGuesses;

        public HangmanGame(String hiddenWord, int maxGuesses) {
            this.word = new Word(hiddenWord);
            this.maxGuesses = maxGuesses;
            this.incorrectGuesses = 0;
        }

        public void play() {
            Scanner scanner = new Scanner(System.in);
            while (!isGameOver() && !isGameWon()) {
                displayCurrentState();
                char guess = getGuess(scanner);
                updateGameState(guess);
            }
            displayCurrentState();
            if (isGameWon()) {
                System.out.println("Congratulations! You've guessed the word!");
            } else {
                System.out.println("Game Over! The word was: " + word.getHiddenWord());
            }
        }

        private void displayCurrentState() {
            System.out.println("Current word: " + word.getGuessedWord());
            System.out.println("Incorrect guesses left: " + (maxGuesses - incorrectGuesses));
        }

        private char getGuess(Scanner scanner) {
            System.out.print("Enter your guess: ");
            return scanner.next().charAt(0);
        }

        private void updateGameState(char guess) {
            if (!word.guessLetter(guess)) {
                incorrectGuesses++;
            }
        }

        private boolean isGameWon() {
            return word.isFullyGuessed();
        }

        private boolean isGameOver() {
            return incorrectGuesses >= maxGuesses;
        }
    }

    public static void main(String[] args) {
        String word = getRandomWord();
        int maxGuesses = 6;
        HangmanGame game = new HangmanGame(word, maxGuesses);
        game.play();
    }
}

File: Hangman.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>
#include <ctime>

using namespace std;

string getRandomWord() {
    vector<string> words = {"example", "hangman", "computer", "programming", "cplusplus", "random", "word"};
    srand(time(0)); 
    int randomIndex = rand() % words.size(); 
    return words[randomIndex]; 
}

bool isAlreadyGuessed(vector<char>& guessedLetters, char letter) {
    for (char guessedLetter : guessedLetters) {
        if (guessedLetter == letter) {
            return true; 
        }
    }
    return false; 
}

class Word {
private:
    string hiddenWord; 
    string guessedWord; 
    vector<char> guessedLetters; 

public:
    Word(string hiddenWord) : hiddenWord(hiddenWord) {
        guessedWord = string(hiddenWord.length(), '_'); 
    }

    bool guessLetter(char letter) {
        if (isAlreadyGuessed(guessedLetters, letter)) {
            return false; 
        }
        guessedLetters.push_back(letter); 

        bool correctGuess = false; 
        for (int i = 0; i < hiddenWord.length(); ++i) {
            if (hiddenWord[i] == letter) {
                guessedWord[i] = letter; 
                correctGuess = true;
            }
        }
        return correctGuess; 
    }

    
    string getGuessedWord() {
        return guessedWord;
    }

    
    string getHiddenWord() {
        return hiddenWord;
    }

    
    bool isFullyGuessed() {
        return guessedWord == hiddenWord;
    }
};


class Hangman {
private:
    Word word; 
    int maxGuesses; 
    int incorrectGuesses; 

public:
    
    Hangman(string hiddenWord, int maxGuesses) : word(hiddenWord), maxGuesses(maxGuesses), incorrectGuesses(0) {}

    
    void play() {
        while (!isGameOver() && !isGameWon()) {
            displayCurrentState();
            char guess = getGuess(); 
            updateGameState(guess); 
        }
        displayCurrentState(); 
        if (isGameWon()) {
            cout << "Congratulations! You've guessed the word!" << endl; 
        } else {
            cout << "Game Over! The word was: " << word.getHiddenWord() << endl; 
        }
    }

private:
    
    void displayCurrentState() {
        cout << "Current word: " << word.getGuessedWord() << endl; 
        cout << "Incorrect guesses left: " << maxGuesses - incorrectGuesses << endl; 
    }

   
    char getGuess() {
        char guess;
        cout << "Enter your guess: ";
        cin >> guess; 
        return guess;
    }

  
    void updateGameState(char guess) {
        if (!word.guessLetter(guess)) {
            incorrectGuesses++; 
        }
    }

    
    bool isGameWon() {
        return word.isFullyGuessed();
    }

    
    bool isGameOver() {
        return incorrectGuesses >= maxGuesses;
    }
};


int main() {
    string word = getRandomWord(); 
    int maxGuesses = 6; 
    Hangman game(word, maxGuesses); 
    game.play(); 
    return 0;
}
